src/api/bulk2.ts

```diff
@@ -0,0 +1,1082 @@
+import { EventEmitter } from 'events';
+import { Duplex, Readable, Writable } from 'stream';
+import Connection from '../connection';
+import { Serializable, Parsable } from '../record-stream';
+import HttpApi from '../http-api';
+import { StreamPromise } from '../util/promise';
+import { registerModule } from '../jsforce';
+import { BulkRequest } from './bulk';
+import { getLogger, Logger } from '../util/logger';
+import { concatStreamsAsDuplex } from '../util/stream';
+import { HttpResponse, Record, Schema, Optional } from '../types';
+import is from '@sindresorhus/is';
+
+export type IngestOperation =
+  | 'insert'
+  | 'update'
+  | 'upsert'
+  | 'delete'
+  | 'hardDelete';
+
+type BaseJobInfo = {
+  id: string;
+  object: string;
+  createdById: string;
+  createdDate: string;
+  systemModstamp: string;
+  apiVersion: number;
+  lineEnding: 'LF' | 'CRLF';
+  columnDelimiter:
+    | 'BACKQUOTE'
+    | 'CARET'
+    | 'COMMA'
+    | 'PIPE'
+    | 'SEMICOLON'
+    | 'TAB';
+  concurrencyMode: 'Parallel';
+  contentType: 'CSV';
+  numberRecordsProcessed: number;
+  retries: number;
+  totalProcessingTime: number;
+};
+
+export type QueryJobInfoV2 = BaseJobInfo & {
+  operation: 'query' | 'queryAll';
+  state: 'UploadComplete' | 'InProgress' | 'Aborted' | 'JobComplete' | 'Failed';
+  jobType: 'V2Query';
+  isPkChunkingSupported: boolean;
+};
+
+export type JobInfoV2 = BaseJobInfo & {
+  apexProcessingTime: number;
+  apiActiveProcessingTime: number;
+  assignmentRuleId?: string;
+  contentUrl: string;
+  errorMessage?: string;
+  externalIdFieldName?: string;
+  jobType: 'BigObjectIngest' | 'Classic' | 'V2Ingest';
+  operation: IngestOperation;
+  state:
+    | 'Open'
+    | 'UploadComplete'
+    | 'InProgress'
+    | 'JobComplete'
+    | 'Aborted'
+    | 'Failed';
+  numberRecordsFailed: number;
+};
+
+export type IngestJobV2SuccessfulResults<S extends Schema> = Array<
+  {
+    sf__Created: 'true' | 'false';
+    sf__Id: string;
+  } & S
+>;
+
+export type IngestJobV2FailedResults<S extends Schema> = Array<
+  {
+    sf__Error: string;
+    sf__Id: string;
+  } & S
+>;
+
+export type IngestJobV2UnprocessedRecords<S extends Schema> = Array<S>;
+
+export type IngestJobV2Results<S extends Schema> = {
+  successfulResults: IngestJobV2SuccessfulResults<S>;
+  failedResults: IngestJobV2FailedResults<S>;
+  unprocessedRecords: IngestJobV2UnprocessedRecords<S>;
+};
+
+type NewIngestJobOptions = Required<Pick<JobInfoV2, 'object' | 'operation'>> &
+  Partial<
+    Pick<
+      JobInfoV2,
+      | 'assignmentRuleId'
+      | 'columnDelimiter'
+      | 'externalIdFieldName'
+      | 'lineEnding'
+      | 'contentType'
+    >
+  >;
+
+type NewQueryJobOptions = {
+  query: string;
+  operation: QueryJobInfoV2['operation'];
+} & Partial<Pick<QueryJobInfoV2, 'columnDelimiter' | 'lineEnding'>>;
+
+type CreateIngestJobV2Options = {
+  bodyParams: NewIngestJobOptions;
```

Comment:
I would have preferred `options` instead of `bodyParams` but this is already inside an `options` object that constructors take.